File: python/custom-set/custom_set.py
"""
Custom Set Class Implementation
"""

class CustomSet:
    """
    Custom Set class
    I will resist the temptation to just route this through a python set.
    A list will be used instead.
    :param elements: Elements to initialize the custom set with
    """
    def __init__(self, elements=None):
        self.storage = []
        if elements is None:
            elements = []
        for element in elements:
            self.add(element)

    def isempty(self):
        """
        Check if the set is empty
        :returns: True if an empty set
        """
        return len(self.storage) == 0

    def __contains__(self, element):
        """
        Check if the custom set contains a specific element
        :param element: What to search for in the set.
        :returns: True if the element is found.
        """
        return element in self.storage

    def issubset(self, other):
        """
        Check if this set is a sub set of the other custom set
        :param other: The set to check is a equal to/superset of this set.
        :returns: True if this set is a subset of the other set.
        """
        missing = [element for element in self.storage if element not in other.storage]
        return len(missing) == 0

    def isdisjoint(self, other):
        """
        Check if this set is disjoint with another set (they have no overlapping items)
        :param other: The set to check against
        :returns: True if both sets are non-empty and contain unique items.
        """
        if self.isempty() or other.isempty():
            return True
        
        shared = [element for element in self.storage if element in other.storage]
        return len(shared) == 0

    def __eq__(self, other):
        """
        Check if two sets are equal
        :param other: The other custom set object to compare against
        :returns: True if the sets are the same.
        """
        # False if they are different lengths.
        if len(self.storage) != len(other.storage):
            return False
        
        # If any element in this set is not in the other they are not equal.
        unequal = [element for element in self.storage if element not in other.storage]        
            
        # Same length, same elements
        return len(unequal) == 0

    def __hash__(self) -> int:
        """
        Return a hash of this custom set object.
        Only added this because exercism complains if you have eq but not hash
        :returns: hash of this custom set object.
        """
        return hash(self.storage)
        

    def add(self, element):
        """
        Add an element into the custom set if it is not already present.
        :param element: The item to add to the custom set
        """
        if not element in self.storage:
            self.storage.append(element)

    def intersection(self, other):
        """
        Return a new set that contains elements common to this set and the other set
        :param other: Custom set to compare against
        :returns: A new custom set object containing the shared elements
        """
        return CustomSet([element for element in self.storage if element in other.storage])

    def __sub__(self, other):
        """
        Return a new set that contains elements in self that are not in the other set.
        :param other: Custom set to compare against
        :returns: A new custom set object containing the elements unique to this set.
        """
        return CustomSet([element for element in self.storage if element not in other.storage])

    def __add__(self, other):
        """
        Return a new set that cointains the unique elements between this set and the other set
        :param other: The Custom set to make a new set with
        :returns: A new custom set with unique elements from both sets
        """
        ret = [*self.storage]
        ret.extend([element for element in other.storage if element not in self.storage])
        return CustomSet(ret)

File: go/custom-set/custom_set.go
// Custom Set Exercise.
package customset

import (
	"fmt"
	"sort"
	"strings"
)

// Implement Set as a collection of unique string values.
//
// For Set.String, use '{' and '}', output elements as double-quoted strings
// safely escaped with Go syntax, and use a comma and a single space between
// elements. For example, a set with 2 elements, "a" and "b", should be formatted as {"a", "b"}.
// Format the empty set as {}.

// Don't think I escaped the strings right.

// Define the Set type here.
// I put it in a struct so the data field is by reference
// which allows us to change it (when sorting or adding)
type Set struct {
	data []string
}

// Required for sorting. Return the length of the set.
// @returns: Number of elements in the set.
func (s Set) Len() int {
	return len(s.data)
}

// Required for sorting. Swap two elements in set by index.
// @param i: First index
// @param j: Second index
func (s Set) Swap(i, j int) {
	s.data[i], s.data[j] = s.data[j], s.data[i]
}

// Required for sorting. Return if value at index i is less than the value at index j
// @param i: First index
// @param j: Second index
// @returns: True if data at i < data at j.
func (s Set) Less(i, j int) bool {
	return s.data[i] < s.data[j]
}

// Create a new empty set
// @returns: The new set.
func New() Set {
	return Set{data: []string{}}
}

// Create a new set with the elements of the given slice.
// @param l: Slice of string to populate the set with.
// @returns: New set object populated with the elements in l.
func NewFromSlice(l []string) Set {
	ret := Set{}
	for i := 0; i < len(l); i++ {
		ret.Add(l[i])
	}
	return ret
}

// Convert a set into a string representation.
// @returns: String representation of the set.
func (s Set) String() string {
	ret := strings.Builder{}
	ret.WriteString("{")
	for i := 0; i < s.Len(); i++ {
		if i > 0 {
			ret.WriteString(", ")
		}
		ret.WriteString(fmt.Sprintf("\"%s\"", s.data[i]))
	}
	ret.WriteString("}")
	return ret.String()
}

// Check to see if a set is empty
// @returns: True if empty.
func (s Set) IsEmpty() bool {
	return s.Len() == 0
}

// Check to see if the set has a specific element.
// @param elem: The value to search for.
// @returns: True if the element was found in the set.
func (s Set) Has(elem string) bool {
	if s.Len() == 0 {
		return false
	}
	return s.BinSearch(elem, 0, s.Len()-1) != -1
}

// Binary Search. We keep the elements in the set sorted for fast lookup
// @param elem: The string to search for.
// @param begin: Beginning of range to search in.
// @param end: End of range to search in.
// @returns: -1 if not found otherwise the current index of the element in the set.
func (s Set) BinSearch(elem string, begin, end int) int {
	// degenerate
	if begin > end {
		return -1
	}

	// One element
	if begin == end {
		if s.data[begin] == elem {
			return begin
		} else {
			return -1
		}
	}
	// has a range.
	middle := begin + int((end-begin)/2)
	if s.data[middle] == elem {
		return middle
	}
	if s.data[middle] > elem {
		return s.BinSearch(elem, begin, middle-1)
	}

	// Not equal or greater, must be less.
	return s.BinSearch(elem, middle+1, end)
}

// Add an element to the set.
// @param elem: text to add to the set (if not already present).
func (s *Set) Add(elem string) {
	if !s.Has(elem) {
		s.data = append(s.data, elem)
		sort.Slice(s.data, s.Less)
	}
}

// Check to see if s1 is a subset of s2
// @param s1: The first set
// @param s2: The second set
// @returns: True if all elements in s1 are also in s2.
func Subset(s1, s2 Set) bool {
	if s1.Len() == 0 {
		return true
	}
	if s2.Len() == 0 && s1.Len() > 0 {
		return false
	}
	return Intersection(s1, s2).Len() == s1.Len()
}

// Check if two sets are disjoint
// @param s1: The first set
// @param s2: The second set
// @returns: True if s1 and s2 have no elements in common.
func Disjoint(s1, s2 Set) bool {
	if s1.Len() == 0 || s2.Len() == 0 {
		return true
	}
	return Intersection(s1, s2).Len() == 0
}

// Check if two sets are equal
// @param s1: The first set
// @param s2: The second set
// @returns: True if s1 is the same length as s2 and has the same elements
func Equal(s1, s2 Set) bool {
	if s1.Len() != s2.Len() {
		return false
	}
	return Intersection(s1, s2).Len() == s1.Len()
}

// Return the intersection of two sets
// @param s1: The first set
// @param s2: The second set
// @returns: A new set with the elements in both s1 and s2.
func Intersection(s1, s2 Set) Set {
	var ret Set = Set{}
	for i := 0; i < s1.Len(); i++ {
		elem := s1.data[i]
		if s2.Has(elem) {
			ret.Add(elem)
		}
	}
	return ret
}

// Return the Difference of two sets
// @param s1: The first set
// @param s2: The second set
// @returns: A set with everything in s1, that is not in s2.
func Difference(s1, s2 Set) Set {
	var ret Set = Set{}
	for i := 0; i < s1.Len(); i++ {
		elem := s1.data[i]
		if !s2.Has(elem) {
			ret.Add(elem)
		}
	}
	return ret
}

// Return the union of two sets.
// @param s1: The first set
// @param s2: The second set
// @returns: A new set with the elements of both sets.
func Union(s1, s2 Set) Set {
	var ret Set = Set{}
	for i := 0; i < s1.Len(); i++ {
		ret.Add(s1.data[i])
	}
	for i := 0; i < s2.Len(); i++ {
		ret.Add(s2.data[i])
	}
	return ret
}
